vissl/data/torchvision_dataset.py

```diff
@@ -0,0 +1,110 @@
+from typing import Tuple, List, Optional, Callable, Any
+
+from PIL import Image
+from fvcore.common.file_io import PathManager
+from torch.utils.data import Dataset
+from torchvision.datasets import CIFAR10, CIFAR100, STL10, MNIST
+
+from vissl.utils.hydra_config import AttrDict
+
+
+class TorchvisionDataset(Dataset):
+    """
+    Adapter dataset to available datasets in Torchvision.
+    The selected dataset is based on the name provided as argument.
+    This name must match the names of the dataset in torchvision.
+
+    Args:
+        cfg (AttrDict): configuration defined by user
+        data_source (string): data source ("torchvision_dataset") [not used]
+        path (string): path to the dataset
+        split (string): specify split for the dataset (either "train" or "val").
+        dataset_name (string): name of dataset.
+    """
+
+    def __init__(self,
+                 cfg: AttrDict,
+                 data_source: str,
+                 path: str,
+                 split: str,
+                 dataset_name: str):
+        super().__init__()
+        assert PathManager.isdir(path), f"Directory {path} does not exist"
+        self.dataset_name = dataset_name
+        self.path = path
+        self.split = split.lower()
+        self.dataset = self._load_dataset()
+
+    def _load_dataset(self):
+        is_train_split = self.split == "train"
+        if self.dataset_name == "CIFAR10":
+            return CIFAR10(self.path, train=is_train_split)
+        elif self.dataset_name == "CIFAR100":
+            return CIFAR100(self.path, train=is_train_split)
+        elif self.dataset_name == "STL10":
+            stl_split = "train" if is_train_split else "test"
+            return STL10(self.path, split=stl_split)
+        elif self.dataset_name == "MNIST":
+            return MNISTAdapter(root=self.path, train=is_train_split)
+        else:
+            raise ValueError(f"Unsupported dataset {self.dataset_name: str}")
+
+    def num_samples(self) -> int:
+        """
+        Size of the dataset
+        """
+        return len(self.dataset)
+
+    def __len__(self) -> int:
+        """
+        Size of the dataset
+        """
+        return self.num_samples()
+
+    def __getitem__(self, idx: int) -> Tuple[Image.Image, bool]:
+        """
+        Return the image at index 'idx' and whether the load was successful
+        Note: we do delayed loading of data to reduce the memory size
+              due to pickling of dataset across dataloader workers.
+        """
+        image = self.dataset[idx][0]
+        is_success = True
+        return image, is_success
+
+    def get_labels(self) -> List[int]:
+        """
+        Return the labels for each sample
+        """
+        return [self.dataset[i][1] for i in range(self.num_samples())]
+
+
+class MNISTAdapter(MNIST):
```

Comment:
is this something that makes sense to implement as a transform for mnist only? that way we can keep the data transformation logic to transformations only. see some transformations like https://github.com/facebookresearch/vissl/blob/master/vissl/data/ssl_transforms/img_pil_to_lab_tensor.py where we convert images to LAB format. I would highly encourage to move this to transform the more I look at it unless it won't fit the bill :)

Reply:
It makes sense as a transform indeed ! I will rework that and get rid of the MNIST adapter which I did not find very satisfying either :)

Reply:
This should be fixed as of 5a86a206e4f29d6c659db314cd55ee1e533c03c4. The code is better that way and can now be tested, which is really great :)